lib/recursive-methods.rb

```diff
@@ -1,49 +1,198 @@
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  # accepts a string and returns the reverse of the string by reversing all letters and all words in the string
+  # ex: reverse("hello world") will convert the input string to "dlrow olleh"
+  if s.length <= 1
+    return s 
+  else
+    return s[-1] + reverse(s[1...-1]) + s[0]
+  end
+  
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(???)
+# Space complexity: O(???)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  #  accepts a string as a parameter and then reverses the string IN PLACE using a recursive algorithm.
+  left_index = 0
+  right_index = s.length - 1
+  
+  swap(s, left_index, right_index)
+  
+  return s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# recursive fcn to be used inside reverse_inplace()
+def swap(s, left_index, right_index)
+  if left_index >= right_index
+    return 
+  else
+    temp = s[left_index] 
+    s[left_index] = s[right_index]
+    s[right_index] = temp
+    left_index += 1
+    right_index -= 1
+    swap(s, left_index, right_index)
+  end
+end
+
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  #   N represents a number of bunnies and each bunny has two big floppy ears. We want to compute the total number of ears across all the bunnies recursively (without loops or multiplication).
+  if n == 0
+    # base case: no bunnies left
+    return 0
+  elsif n >= 1
+    # recurse
+    return 2 + bunny(n-1)
+  else
+    # no negative bunnies
+    raise ArgumentError
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  # puts "\nTESTING ON #{s}"
+  
+  # read from front until ( is found, return false if ) instead
+  i = 0
+  left_paren_found = false
+  while s[i] && left_paren_found==false
+    if s[i] == ")"
+      # puts "NOPE"
+      return false
+    elsif s[i] != "("
+      i += 1
+    else  
+      # puts "\tfound ( at index #{i}"
+      left_paren_found = true
+    end
+  end
+  
+  # read from back until ) is found, return false if ( instead
+  j = s.length-1
+  right_paren_found = false
+  while s[j] && right_paren_found==false
+    if s[j] == "("
+      # puts "BUSTED"
+      return false
+    elsif s[j] != ")"
+      j -= 1
+    else  
+      # puts "\tfound ) at index #{j}"
+      right_paren_found = true
+    end
+  end
+  
+  if (right_paren_found && left_paren_found) && (i < j)
+    # puts "recursion..."
+    return nested(s[i+1...j])
+  elsif !right_paren_found && !left_paren_found
+    return true
+  else
+    return false
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
```

Comment:
1.  You are doing recursion so the system stack applies to the space complexity (at least O(n))
1.  Since you are doing `array.shift` this is by itself an O(n) time operation.  Combined with the recursion it's O(n<sup>2</sup>)